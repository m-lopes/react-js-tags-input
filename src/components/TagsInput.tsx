import React from 'react'

import Box from '@material-ui/core/Box'
import Button from '@material-ui/core/Button'
import Chip from '@material-ui/core/Chip'
import Grid from '@material-ui/core/Grid'
import InputAdornment from '@material-ui/core/InputAdornment'
import TextField from '@material-ui/core/TextField'

import { createStyles, makeStyles, Theme } from '@material-ui/core'
import { ChangeEvent, KeyboardEvent, useEffect, useState } from 'react'

import { getById, update } from '@/http/mailingListsRequests'
import { MailingList } from '@/models/mailingLists'
import { multipleTags } from '@/utils/multipleTags'
import { tagAlreadyExists } from '@/utils/tagAlreadyExists'
import { verifyingValidEmail } from '@/utils/verifyingValidEmail'

interface TagsInputProps {
  tags?: string[]
}

export function TagsInput({ tags }: TagsInputProps) {
  const [inputTags, setInputTags] = useState('')
  const [emailsTags, setEmailsTags] = useState([] as string[])
  const [mailingList, setMailingList] = useState({} as MailingList)

  const css = makeStyles((theme: Theme) =>
    createStyles({
      tagsWrapper: {
        marginTop: theme.spacing(3),
        display: 'flex',
        justifyContent: 'center',
        flexWrap: 'wrap',
        '& > *': {
          margin: theme.spacing(0.5),
        },
      },
    })
  )()

  useEffect(() => {
    async function getData() {
      const { data } = await getById('3ba1d46f-cdef-aafb-f67e-a0f3bd4b1473')

      setMailingList({ ...data, emails: [...tags, ...data.emails] })
      setEmailsTags([...tags, ...data.emails])
    }

    getData()
  }, [tags])

  // verificar se a tag é um email válido e se ela já existe
  function itIsPossibleToAddTheTag(tag: string) {
    return verifyingValidEmail(tag) && !tagAlreadyExists(emailsTags, tag)
  }

  // verificar se a tag já está sendo considerada quando múltipas tags são inseridas
  function theTagHasAlreadyBeenConsidered(consideredTags: string[], newTag: string) {
    return tagAlreadyExists(consideredTags, newTag)
  }

  // adicionar tags
  function onAddNewTags() {
    if (multipleTags(inputTags)) {
      const typedTags = inputTags.split(';')
      const tagsPrepared: string[] = []

      typedTags.forEach(tag => {
        if (itIsPossibleToAddTheTag(tag) && !theTagHasAlreadyBeenConsidered(tagsPrepared, tag)) {
          tagsPrepared.push(tag)
        }
      })

      setEmailsTags([...emailsTags, ...tagsPrepared])
      setInputTags('')
    } else {
      const newTag = inputTags

      if (verifyingValidEmail(newTag) && !tagAlreadyExists(emailsTags, newTag)) {
        setEmailsTags([...emailsTags, newTag])
        setInputTags('')
      } else {
        setInputTags('')
      }
    }
  }

  // deletar a última tag
  function onDeleteTheLastTag() {
    const deletedTag = emailsTags.pop()

    setEmailsTags(emailsTags.filter(tag => tag !== deletedTag))
  }

  // deletar uma tag
  function handleDeleteTag(tagToDelete: string) {
    setEmailsTags(emailsTags.filter(tag => tag !== tagToDelete))
  }

  // pegar as tags digitadas
  function handleChangeInputTags(event: ChangeEvent<HTMLInputElement>) {
    setInputTags(event.target.value)
  }

  // decidir ação com base na tecla precionada pelo usuário
  function handleKeyDown(event: KeyboardEvent) {
    switch (event.key) {
      case 'Tab':
        onAddNewTags()
        break

      case 'Enter':
        onAddNewTags()
        break

      case 'Backspace':
        onDeleteTheLastTag()
        break

      default:
        break
    }
  }

  // atualizar a mailing list
  async function handleSubmit() {
    const { success, data } = await update({ ...mailingList, emails: emailsTags })

    if (success) {
      setMailingList(data)
      setEmailsTags([...tags, ...data.emails])
    }
  }

  return (
    <>
      <Grid container justify="flex-end" style={{ marginBottom: 20 }}>
        <Grid item>
          <Button size="small" variant="contained" color="primary" onClick={handleSubmit}>
            SUBMIT
          </Button>
        </Grid>
      </Grid>

      <Grid container justify="center">
        <Grid item>
          <TextField
            fullWidth
            id="tags"
            name="tags"
            margin="dense"
            placeholder="add tags"
            value={inputTags}
            label="tags"
            variant="outlined"
            onKeyDown={handleKeyDown}
            onChange={handleChangeInputTags}
            inputProps={{ 'data-testid': 'tags' }}
            InputProps={{
              startAdornment: <InputAdornment position="start">To:</InputAdornment>,
            }}
          />

          <Box className={css.tagsWrapper}>
            {emailsTags.map(tag => (
              <Chip
                key={tag}
                label={tag}
                size="small"
                tabIndex={-1}
                data-testid={tag}
                onDelete={() => handleDeleteTag(tag)}
              />
            ))}
          </Box>
        </Grid>
      </Grid>
    </>
  )
}

TagsInput.defaultProps = {
  tags: [],
}
