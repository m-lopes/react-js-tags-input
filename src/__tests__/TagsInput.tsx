import React from 'react'
import { cleanup, fireEvent, render, screen } from '@testing-library/react'

import { TagsInput } from '@/components/TagsInput'

describe('TagsInput Component', () => {
  afterEach(jest.clearAllMocks)
  afterEach(cleanup)

  it('cria o snapshot do component', () => {
    const container = render(<TagsInput />)
    expect(container.asFragment()).toMatchSnapshot()
  })

  it('deve renderizar as tags enviadas por atributos', () => {
    const emails = ['contato@rarolabs.com.br', 'nao-responda@rarolabs.com.br']

    const { debug } = render(<TagsInput tags={emails} />)

    debug()
  })

  it('deve renderizar tags quando preencher o input e pressionar enter', () => {
    render(<TagsInput />)

    const input = screen.getByTestId('tags')

    fireEvent.change(input, { target: { value: 'contato@rarolabs.com.br' } })
    fireEvent.keyDown(input, {
      key: 'Enter',
      code: 'Enter',
      keyCode: 13,
      charCode: 13,
    })

    expect(screen.getByTestId('contato@rarolabs.com.br')).toBeTruthy()
  })

  it('deve renderizar tags quando preencher o input e pressionar tab', () => {
    render(<TagsInput />)

    const input = screen.getByTestId('tags')

    fireEvent.change(input, { target: { value: 'contato@rarolabs.com.br' } })
    fireEvent.keyDown(input, {
      key: 'Tab',
      code: 'Tab',
      keyCode: 9,
      charCode: 9,
    })

    expect(screen.getByTestId('contato@rarolabs.com.br')).toBeTruthy()
  })

  it('deve deletar a útima tag criada ao pressionar o botão de backspace', async () => {
    render(<TagsInput />)

    const input = screen.getByTestId('tags')

    fireEvent.change(input, { target: { value: 'contato@rarolabs.com.br' } })
    fireEvent.keyDown(input, {
      key: 'Tab',
      code: 'Tab',
      keyCode: 9,
      charCode: 9,
    })

    expect(screen.getByTestId('contato@rarolabs.com.br')).toBeTruthy()

    fireEvent.keyDown(input, {
      key: 'Backspace',
      code: 'Backspace',
      keyCode: 8,
      charCode: 8,
    })

    expect(screen.queryByTestId('contato@rarolabs.com.br')).toBeFalsy()
  })
})
