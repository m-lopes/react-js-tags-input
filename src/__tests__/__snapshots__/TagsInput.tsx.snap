// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsInput Component cria o snapshot do component 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-flex-end"
    style="margin-bottom: 20px;"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          SUBMIT
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense MuiInputLabel-outlined"
          data-shrink="true"
          for="tags"
        >
          tags
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-marginDense MuiOutlinedInput-marginDense"
        >
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PrivateNotchedOutline-legend-3"
              style="width: 0px;"
            >
              <span>
                â€‹
              </span>
            </legend>
          </fieldset>
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-marginDense"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              To:
            </p>
          </div>
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
            data-testid="tags"
            id="tags"
            name="tags"
            placeholder="add tags"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-4 makeStyles-tagsWrapper-1"
      />
    </div>
  </div>
</DocumentFragment>
`;
